### PR TITLE
Integrate Blockchain and Wallet model and services with new transaction processing and multi wallet support 

### PR DESCRIPTION
**Pull request by mod**

### CODE REVIEW

File: config/templates/seed/blockchains.yml.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  server:           http://127.0.0.1:8545
+  height:           2500000
+  explorer:
+    address:        https://etherscan.io/address/#{address}
+    transaction:    https://etherscan.io/tx/#{txid}
+  status:           active
+
+- key:              eth-mainet
+  name:             Ethereum Mainet
+  client:           eth
```
Comment (ysv) [inline config/templates/seed/blockchains.yml.erb]: Don't we have one to one relation between blockchain and currency ?

**Comment by ysv** [comment]
Please add blank lines in seeds.rb and seeds.rake

File: config/templates/seed/blockchains.yml.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+- key:              eth-rinkeby
+  name:             Ethereum Rinkeby
+  client:           eth
+  server:           http://127.0.0.1:8545
```
Comment (ysv) [inline config/templates/seed/blockchains.yml.erb]: Will blockchain model support only crypto nodes ? (without bitgo)

File: config/templates/seed/blockchains.yml.erb (hunk 1 above)
Comment (mod) [reply config/templates/seed/blockchains.yml.erb]: yes currency will have a blockchain-key field
it's TODO

File: config/templates/seed/blockchains.yml.erb (hunk 2 above)
Comment (mod) [reply config/templates/seed/blockchains.yml.erb]: no this also work with bitgo but you need a full-node to get database from
you can use both a cryptonode and make payment using bitgo

File: config/templates/seed/blockchains.yml.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  explorer:
+    address:        https://etherscan.io/address/#{address}
+    transaction:    https://etherscan.io/tx/#{txid}
+  status:           active
+
+- key:              eth-mainet
+  name:             Ethereum Mainet
+  client:           eth
+  server:           http://127.0.0.1:8545
+  height:           5000000
```
Comment (calj) [inline config/templates/seed/blockchains.yml.erb]: we need to keep the height of the blockchain to compute confirmations count of transactions, and the pointer to the latest scanned block. For example geth API call eth_syncing returns the following:
- startingBlock: QUANTITY - The block at which the import started (will only be reset, after the sync reached his head)
- currentBlock: QUANTITY - The current block, same as eth_blockNumber
- highestBlock: QUANTITY - The estimated highest block

File: db/migrate/20180708014826_create_blockchains.rb [hunk 4]
```diff
@@ -1,17 +1,18 @@
 class CreateBlockchains < ActiveRecord::Migration
   def change
     create_table :blockchains do |t|
-      t.string :key
-      t.string :name
-      t.string :client
-      t.string :server
+      t.string  :key,                 null: false, index: { unique: true }
+      t.string  :status,              index: true
```
Comment (mod) [inline db/migrate/20180708014826_create_blockchains.rb]: leave state field before created_at
timestamps should always be the last fields.

File: db/schema.rb [hunk 5]
```diff
@@ -80,6 +81,7 @@
     t.integer  "precision",            limit: 1,                              default: 8,      null: false
     t.datetime "created_at",                                                                   null: false
     t.datetime "updated_at",                                                                   null: false
+    t.string   "blockchain_key",       limit: 32
```
Comment (mod) [inline db/schema.rb]: move at the beginning of the table

File: lib/block_api.rb [hunk 6]
```diff
@@ -98,7 +96,7 @@ def inspect_address!(address)
     end
 
     def convert_to_base_unit!(value)
-      x = value.to_d * currency.base_factor
+      x = value.to_d * blockchain.base_factor
```
Comment (ysv) [inline lib/block_api.rb]: Base factor is currency specific

**Comment by ysv** [comment]
Use this [scrip](https://gist.github.com/ysv/d249b47adf8f3b25752b74b3ad205772) to fetch information about bunch of blocks in ethereum
Usage 
```./myloop.sh 2621839 2621843 > data.json```
2621839 - first block
2621843 - last block

**Comment by Coding-Jewelry** [comment]
Hi @ysv , hope you are doing well today.
When do you think it will be finished?
I want to remove the ETH deposit/withdrawal issue as soon as possible.
So I am waiting for your update.
Thank you.

PS:  Which branch are you working on currently?

**Comment by ysv** [comment]
Hi @Coding-Jewelry as you can see we've just finished with ERC20 and ETH implementation https://github.com/rubykube/peatio/pull/1427#event-1740801341
But also we need to implement same services for other currencies and BitGo so it will take some time.
You are welcome to use this branch for local testing of ERC20.

**Comment by Coding-Jewelry** [comment]
Hi @ysv , thank you for your updates.
So the withdrawal issue that I mentioned in https://github.com/rubykube/peatio/issues/1416, has also gone away?

**Comment by ysv** [comment]
@Coding-Jewelry yes once we will implement multiple wallet support

**Comment by Coding-Jewelry** [comment]
Hi @ysv ,  I was asking about the depositing/withdrawing issue of ETH/ERC20 tokens.  So is this branch working well for now? For depsits/withdrawals of ETH/ERC20?  Thank you. :)

**Comment by Coding-Jewelry** [comment]
Hi @ysv , hope you are doing well today.
In seed/wallets.yml file, how can I understand "Ethereum Deposit Wallet"?  What's this and what's the difference between this and "Ethereum Hot Wallet"?
And can I set a smart contract address as Ethereum Hot Wallet's address?
Thank you.

**Comment by ysv** [comment]
@Coding-Jewelry We are going to support separate wallets for deposits and withdrawals
All moneys in deposit wallet will be transferred to withdraw wallets (hot, warm or cold)
You can replace coins and deposit_coin daemon with blockchain daemon

**Comment by Coding-Jewelry** [comment]
Hi @ysv , could you please explain about this in detail:
"You can replace coins and deposit_coin daemon with blockchain daemon"

File: lib/client/bitcoin.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    end
+
+    def build_transaction(tx, current_block, latest_block, address)
+      entries = tx.fetch('vout').map do |item|
+
+        next if item.fetch('value').to_d <= 0
+        next unless item['scriptPubKey'].has_key?('addresses')
+        next if address != normalize_address(item['scriptPubKey']['addresses'][0])
+
+        { amount: item.fetch('value').to_d, address: normalize_address(item['scriptPubKey']['addresses'][0]) }
```
Comment (dinesh-skyach) [inline lib/client/bitcoin.rb]: @mod I'm not sure if this is the correct way.

File: app/models/deposits/fiat.rb [hunk 8]
```diff
@@ -12,25 +12,25 @@ def charge!
...
+#  id           :integer          not null, primary key
+#  member_id    :integer          not null
+#  currency_id  :string(10)       not null
+#  amount       :decimal(32, 16)  not null
+#  fee          :decimal(32, 16)  not null
+#  address      :string(64)
+#  txid         :string(128)
+#  txout        :integer
+#  aasm_state   :string(30)       not null
+#  block_number :integer
```
Comment (mod) [inline app/models/deposits/fiat.rb:28]: @dinesh-skyach do we also save the block_number in which we found withdrawal ?

File: app/models/withdraws/coin.rb [hunk 9]
```diff
@@ -21,38 +21,57 @@ class Coin < Withdraw
...
-      end
+    def confirmations
+      return 0 if block_number.blank?
+      latest_block_number - block_number
+    rescue Faraday::ConnectionFailed => e
+      report_exception(e)
+      'N/A'
+    end
+
+    # TODO: backport audit!
```
Comment (mod) [inline app/models/withdraws/coin.rb:47]: @dmk can you look into it?

File: app/services/blockchain_service.rb [hunk 10]
```diff
@@ -0,0 +1,97 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module BlockchainService
+  Error                  = Class.new(StandardError) # TODO: Rename to Exception.
+  ConnectionRefusedError = Class.new(StandardError) # TODO: Remove this.
```
Comment (mod) [inline app/services/blockchain_service.rb:6]: Please define more errors and rename them

File: app/services/blockchain_service.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    def update_or_create_deposits!(deposits)
+      deposits.each do |deposit_hash|
+
+        # If deposit doesn't exist create it.
+        deposit = Deposits::Coin
+                    .where(currency: currencies)
+                    .find_or_create_by!(deposit_hash)
+
+        deposit.accept! if deposit.confirmations >= blockchain.min_confirmations
```
Comment (mod) [inline app/services/blockchain_service.rb:41]: @dinesh-skyach what is the update part of it?

File: app/services/wallet_service.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module WalletService
+  Error                  = Class.new(StandardError) # TODO: Rename to Exception.
+  ConnectionRefusedError = Class.new(StandardError) # TODO: Remove this.
```
Comment (mod) [inline app/services/wallet_service.rb:6]: We need more detailed errors

**Review by mod — COMMENTED** [review]
@dinesh-skyach can you review the branch, rebase on master, and squash migrations into 1 create table per table

File: app/models/deposits/fiat.rb (hunk 8 above)
Comment (dinesh-skyach) [reply app/models/deposits/fiat.rb:28]: @mod Yes, we do it here https://github.com/rubykube/peatio/blob/feature/blockchains/app/services/blockchain_service.rb#L61

File: app/services/blockchain_service.rb (hunk 11 above)
Comment (dinesh-skyach) [reply app/services/blockchain_service.rb:41]: @mod we update the state of deposit if min_confirmations are matched.

File: lib/wallet_client/bitgo.rb [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    def initialize(*)
+      super
+      currency_code_prefix = wallet.bitgo_test_net ? 't' : ''
+      @endpoint            = wallet.bitgo_rest_api_root.gsub(/\/+\z/, '') + '/' + currency_code_prefix + wallet.currency.code
+      @access_token        = wallet.bitgo_rest_api_access_token
+    end
+
+    def load_balance!
+      convert_from_base_unit(wallet_details(true).fetch('balanceString'))
```
Comment (sramsden) [inline lib/wallet_client/bitgo.rb:15]: ☝️  BUG! .. `wallet_details` method below 👇 does not take any parameters.

**Pull request merged**